src/gazebo_vision_plugin.cpp

```diff
@@ -0,0 +1,173 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+ /**
+  * @brief Vision Plugin
+  *
+  * This plugin simulates VIO data
+  *
+  * @author Christoph Tobler <christoph@px4.io>
+  */
+
+#include <gazebo_vision_plugin.h>
+
+namespace gazebo {
+GZ_REGISTER_MODEL_PLUGIN(VisionPlugin)
+
+VisionPlugin::VisionPlugin() : ModelPlugin()
+{
+}
+
+VisionPlugin::~VisionPlugin()
+{
+  event::Events::DisconnectWorldUpdateBegin(_updateConnection);
+}
+
+void VisionPlugin::Load(physics::ModelPtr model, sdf::ElementPtr sdf)
+{
+  getSdfParams(sdf);
+
+  // Store model
+  _model = model;
+
+  _world = _model->GetWorld();
+  _last_time = _world->GetSimTime();
+  _last_pub_time = _world->GetSimTime();
+
+  // remember start pose -> VIO should always start with zero
+  _pose_model_start = _model->GetWorldPose();
+
+  _nh = transport::NodePtr(new transport::Node());
+  _nh->Init(_namespace);
+
+  // Listen to the update event. This event is broadcast every simulation iteration.
+  _updateConnection = event::Events::ConnectWorldUpdateBegin(
+      boost::bind(&VisionPlugin::OnUpdate, this, _1));
+
+  _pub_odom = _nh->Advertise<odom_msgs::msgs::odom>("~/" + _model->GetName() + "/vision_odom", 10);
+}
+
+float _corellation_time;
+float _random_walk;
+float _noise_density;
+
+void VisionPlugin::OnUpdate(const common::UpdateInfo&)
+{
+  common::Time current_time = _world->GetSimTime();
+  double dt = (current_time - _last_pub_time).Double();
+
+  if (dt > 1.0 / _pub_rate) {
+
+    // get pose of the model that the plugin is attached to
+    math::Pose pose_model_world = _model->GetWorldPose();
+    math::Pose pose_model; // body frame
+    // convert to body frame (front-right-down)
```

Comment:
@ChristophTobler while the data is still on the Gazebo side, this should be propagated in forward-left-up. Then when it is received on the `gazebo_mavlink_interface`, it is rotated to front-right-fown frame.

Reply:
Also, we should have a parameter to decide if the vision position we want is in the body or in the world frame. Initially this was on the world frame.